[Infra/#11] 무중단 배포 자동화 환경 구축 (Docker)

yummygyudon opened the pull request:
<!--
name: Makers Auth Feature PR template
about: 구현한 기능과 변경 사항에 대해 구체적으로 작성해주세요
title: '[FEAT/{#이슈번호}] 기능 내용'
labels: ''
assignees: ''
-->

<!--
- 리뷰어 추가했나요?
- 허가자 추가했나요?
- 라벨 추가했나요?
-->

## Related Issue 🚀
- closed #11 

## Work Description ✏️
- PR 시, 빌드 통관 여부 확인 자동화 환경을 구축했습니다.
- 배포 자동화 환경을 구축했습니다.
- 무중단 배포 환경을 구축했습니다. (노션 팀워크스페이스를 통해 원리와 흐름을 참고해주세요!)

🚨 (주의) **Operation 프로젝트까지** 반영한 후, 진행해야 합니다. - (완료 ✅)
🚨 (주의) 클라이언트와 논의하여 기존 방식으로 구동 중인 어플리케이션들을 종료하고 진행하기 때문에 잠깐의 셧다운이 필요합니다. - (완료 ✅)

<br/>

## PR Point 📸
<!-- 피드백을 받고 싶은 부분을 적어주세요. -->

> to-do
- [x] Github Secrets 반영 필요
- [x] Spring Actuator - 모니터링 서버 스펙 반영 필요

<br/>

### ※ 환경 변수 주입 방식
> `.env` 주입 

기존 각 profile 별 `application-{profile}.yml` 선언 및 import 하는 파일들마다 profile에 맞춰 주입되는 환경변수들을 단일 `.properties` 파일에 선언해야 하는 구조였습니다.

< **as-is** >
- profile 별 환경 변수 정의 (ex. `PROD_A_VALUE` & `DEV_A_VALUE` 중복 선언)
- Docker Build & Run 시, profile 동적 주입의 어려움
- 특정 profile 마다 필요하지 않은 타 Profile에서의 환경변수까지 포함되어 실행됨

< **to-be** >
- Docker 실행 환경에서 간편하게 **다수의 환경변수 일괄 적용 가능**
- Docker 실행 환경에서 간편하게 **실행 profile 지정 가능** (`.env` 파일 내부에 `SPRING_PROFILES_ACTIVE` 선언을 통해 적용)
- 중복되는 profile 별 환경변수를 정의하는 구조 개선
- 각 profile에 맞춰 필요한 환경변수 값들로만 구성된 `.env` 파일들로 분리 가능

(여러분들이 개발하면서 직접 프로젝트 폴더 내에 자유롭게 생성하고 활용할 수 있도록 `.gitignore`에 `**/src/main/resources/**/*.env`를 추가했습니다. 패키지 만드셔서 env 파일 생성하시구 실행 Configuration에서 env 주입하시는 방식으로 실행시키시면 됩니다!)


<br/>

### ※ Build 방식
> build profile parameter(`-Pprofile`) 주입 방식 도입 & CI 시, Docker 환경에서의 Build 방식 변경

< **as-is** >
- ci 전용 properties를 따로 Github Secrets에 정의해놓아야 함
  - 실질적인 값이 필요한 것이 아닌 "Build 성공 여부"만 확인하는 과정이기 때문에 해당 값들은 불필요합니다.
- 특정 profile 마다 필요하지 않은 타 Profile에서의 환경변수까지 포함되어 실행됨

< **to-be** >
- 별도 파라미터 전달이 없을 경우, 기본값으로 "test"로 돌아가도록 구축 (참고 : `build.gradle.kts`)
  - 기본적으로 `application-test.yml` 및 각 import yml 내 test profile 시 사용될 임시 값은 static 하게 정의되어 있어 CI 과정에서 별도 `.env` / `.properties` 파일 주입이 불필요하게 됩니다.
  - dev와 prod 각 환경에서 명시적인 build 환경 주입을 통해 휴먼에러를 방지할 수 있습니다.
- 빌드에 사용되는 `Dockerfile` 내 `ARG` 활용
  - 별도 `--build-arg` 명령이 없을 경우, 기본값으로 "test"로 돌아가도록 구축했습니다. (참고 : `Dockerfile`)


<br/>

### ※ Run 방식
> 위 내용들을 반영하여 환경 별 동적 profile 적용

```
# Example
sudo docker run --name ${Container Name} \
  --env-file application.env \ # profile 환경 변수 포함
  -d -p 8080:8080 \
  -t ${Image Name}:${Tag} # profile 적용 Build Image
```

<br/>

### ※ Deploy 방식
> - CD Script : Copy Resources from S3
> - CodeDeploy : S3 Resource(`.pem`, `.p8`, `.env`, `scripts`) -> EC2 배포 / EC2 내에서 Docker Pull & Run

CodeDeploy에서 실행하는 배포 단계는 아래와 같습니다. ("**무중단 배포 :: Blue-Green**")

1. _Resource Files(S3 Zip) =(Copy)=> EC2_
2. `deploy` script
    1. 기존 구동 컨테이너 확인 
    2. 새로운 이미지 Pull & Run 
    3. 새로운 구동 컨테이너 정상 작동 여부 확인(Health Check)
    4. 기존 구동 컨테이너 종료 및 삭제
3. `switch` script
    1. 새로운 구동 컨테이너 확인
    2. 기존 Nginx 컨테이너 종료 및 삭제
    3. 새로운 구동 컨테이너 전용 Nginx Config 파일로 Volume Mount 경로 변경
    4. 새로운 Nginx 컨테이너 실행
4. `logging` script
    1. 1번 & 2번 완료 시 실행
    2. 배포 일시 로깅


<br/>

### ※ 파일 관리 방식
- **Nginx 관련** 파일 관리
단일 서버 내에서 2개의 어플리케이션을 운영하기 때문에 특정 레포지토리에서 관리하는 구조를 피했습니다.
번거로울순 있지만 "변경이 빈번하지 않다는 점" & "공통 관심사라는 점"을 근거로 해당 관련 파일들(`*.conf`)들은 EC2 내부에서 직접 관리할 수 있도록 했습니다.
(이에 대한 내용들은 노션을 통해 공유할 수 있도록 합니다)

- **실 리소스 파일 관리**
Github Secrets 으로만 관리할 경우, 적용된 값이 올바른 값인지 확인하기 어렵다는 단점이 있습니다.
이에 확인이 필요하거나 직접 관리가 필요한 파일들에 대해서 `S3`에 저장하여 CD 과정에서 Copy하여 사용될 수 있도록 했습니다.
(리소스 파일의 변경이 필요하거나 파일 내용 확인이 필요할 때, S3에서 확인 및 교체가 가능합니다.)

🚨 단, s3에서 관리하는 만큼 **Bucket 이름** & **Key** (경로 및 파일 명)가 노출되지 않도록 해당 값들에 대해서는 Github Secrets에서 관리할 수 있도록 합니다.

yummygyudon committed:
[FEAT] 기본 어플리케이션 빌드 확인 CI Script 작성

yummygyudon committed:
[FEAT] 1차 CI/CD workflow script 작성

yummygyudon committed:
[FEAT] Build 및 Run 도커파일 작성

yummygyudon committed:
[FEAT] Code Deploy 실행 단계 및 실행위치 설정

yummygyudon committed:
[FEAT] Code Deploy `appspec.yml` 실행 Script 순서 지정 및 이름 지정

yummygyudon committed:
[FEAT] CD Script 작성

도커 이미지 Push와 Deploy 파일 세팅 단계를 분리했습니다.

yummygyudon committed:
[FEAT] CI Script 내 Key Pem 파일 생성 명령어 추가

yummygyudon committed:
[FEAT] scripts 환경변수 주입 세팅

yummygyudon committed:
[CHORE] `.gitignore` 내 배포 과정에 사용되는 script 폴더 path 추가

본래 Script는 Github Actions CD 스크립트 실행 시점에 생성되고 S3를 통해 관리되는 자원이지만 작업자들이 프로젝트 작업 과정에서 용이하게 확인하며 관리할 수 있도록 버전관리는 되지 않도록 제외했습니다.

yummygyudon committed:
[CHORE] Spring Actuator 라이브러리 의존성 추가

- Deploy 시 사용되는 Health Check API 제공
- Prometheus 모니터링 지표 Collector 출처로 사용

yummygyudon committed:
[FEAT] 1차 배포 Script 작성 완료

yummygyudon committed:
[FIX] Jwt Pem 파일 환경별 분리에 따른 Script 환경 변수 반영

yummygyudon committed:
[FEAT] Apple Credential 파일 반영

yummygyudon committed:
[FEAT] EC2 내 파일 존재 시, 덮어쓸 수 있도록 appspec.yml 반영

yummygyudon committed:
[FIX] appspec file path 수정

hyunw9 commented on appspec.yml:
```diff
@@ -0,0 +1,26 @@
+version: 0.0
+os: linux
+
+files:
+  - source: /
+    destination: /home/ubuntu
+file_exists_behavior: OVERWRITE
+
+permissions:
+  - object: /home/ubuntu
+    pattern: '**'
+    owner: ubuntu
+    group: ubuntu
+
+hooks:
+  AfterInstall:
+    - location: script/auth/deploy.sh
+      timeout: 120
+      runas: root
+    - location: script/auth/switch.sh
+      timeout: 120
+      runas: root
+  ValidateService:
+    - location: script/auth/valid.sh
+      timeout: 60
+      runas: root
```
이 파일은 무엇을 위한 파일인가요..?

yummygyudon replied:
Code Deploy를 사용하려면 작성해주어야 하는 파일입니다!
자세한 내용은 [공식문서](https://docs.aws.amazon.com/ko_kr/codedeploy/latest/userguide/reference-appspec-file.html)를!

yummygyudon committed:
[FEAT] .dockerignore 추가

yummygyudon committed:
[CHORE] 불필요한 명령 제거 및 브랜치 이름 수정

yummygyudon committed:
[FIX] CodeDeploy resource mapped path 수정

yummygyudon committed:
[FIX] 설정 파일 관리방법 변경 반영

yummygyudon committed:
[CHORE] 배포 변경 사항 Prod script sync

yummygyudon committed:
[FIX] 잘못된 git ignore 등록 요소 수정

yummygyudon committed:
[FIX] CI 빌드 실행 명령 수정

- pem 키등 외부 S3에 관리되고 있는 자원이 필요하여 테스트 실행 없이 빌드되도록 변경했습니다.
- 추후 Test 환경 변수 주입으로 별도 테스트 통과 STEP 생성을 고려해도 좋을 것 같습니다.

yummygyudon committed:
[FIX] appspec 소스파일 저장 위치 변경

yummygyudon committed:
[FEAT] Spring Actuator Health Check 엔드포인트 설정 추가

yummygyudon committed:
[FEAT] appspec pem 파일 및 key 파일 리소스 path 추가

- docker 실행에 필요한 자원들을 추가합니다.
- Github Secret으로 관리되는 자원에 대해 올바르게 주입되었는지 확인하기 위해서도 사용된 리소스를 /home/ubuntu 내에 저장될 수 있도록 합니다.

yummygyudon committed:
[FIX] .env 파일 주입 방식 변경에 대한 application yaml 파일 반영

- 불필요한 properties 선언을 줄이고 env 파일 분리를 통해 환경변수 관리를 용이하게 하고자 했습니다.

yummygyudon committed:
[CHORE] gitignore .env 파일 추가

yummygyudon committed:
[FEAT] 동적 Profile parameter 주입 Build 방식 & 생성 jar 파일 지정 내용 `build.gradle.kts` 반영

- 기본 profile로 test가 주입되도록 설정했습니다.
- Docker Build 및 Run 시, 복수의 jar 파일 추적으로 인해 인식이 안되는 문제를 해결하기 위해 bootJar 시 생성되는 jar 파일의 이름을 직접 지정했습니다.

yummygyudon committed:
[REFACTOR] copy 시, 명시된 jar만 복사할 수 있도록 수정 및 build 주체 변경(Gradle -> OpenJDK)

yummygyudon committed:
[REFACTOR] CI Build 방식 Docker 변경

yummygyudon committed:
[REFACTOR] CD Script 수정

- docker 실행 시, env 파일 주입 방식 반영
- build profile 동적 주입 방식 반영
- 필요한 resource S3 copy 방식 반영

yummygyudon committed:
[FIX] docker 빌드 시, 수행되는 `.github/script/pre-commit` 파일 .dockerignore 제외

sung-silver reviewed (approved):
어렵고 힘들었던 배포네요 수고 정말 많으셨습니다!

yummygyudon committed:
Merge branch 'dev' into infra/#11

yummygyudon committed:
[CHORE] Spotless 스펙 통일
